processing/src/main/java/org/apache/druid/segment/ObjectBasedColumnSelector.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import org.apache.druid.error.DruidException;
+
+/**
+ * Object based column selector.
+ *
+ * This abstract class provides a {@link ColumnValueSelector} based on the methods of {@link BaseObjectColumnValueSelector}.
+ */
+public abstract class ObjectBasedColumnSelector<T> implements ColumnValueSelector<T>
+{
+  private static final String COLUMN_IS_NULL_ERROR_MESSAGE = "Invalid usage pattern: method returning primitive called - but the column is null";
+
+  @Override
+  public final float getFloat()
+  {
+    T value = getObject();
+    if (value == null) {
+      throw DruidException.defensive(COLUMN_IS_NULL_ERROR_MESSAGE);
+    }
+    return ((Number) value).floatValue();
+  }
+
+  @Override
+  public final double getDouble()
+  {
+    T value = getObject();
+    if (value == null) {
+      throw DruidException.defensive(COLUMN_IS_NULL_ERROR_MESSAGE);
+    }
+    return ((Number) value).doubleValue();
+  }
+
+  @Override
+  public final long getLong()
+  {
+    T value = getObject();
+    if (value == null) {
+      throw DruidException.defensive(COLUMN_IS_NULL_ERROR_MESSAGE);
+    }
+    return ((Number) value).longValue();
+  }
+
+  @Override
+  public final boolean isNull()
+  {
+    return getObject() == null;
```

Comment:
I think this should also probably be checking that the result of `getObject()` is `instanceof Number`. The things that are going to call `isNull()` will do so in service of calling it prior to calling `getLong`/`getFloat`/`getDouble`, but if Object is not a `Number` then this will result in a class cast exception and there is no restriction on this class that the things it provides are always instances of a `Number`.

Things that are calling `getObject` should already know not to be checking `isNull` for null checks, so I don't think that would cause any problems if `isNull` returns true but `getObject` returns a non-null value.

Reply:
I'm not sure about the added value of such a check as the classcastexception will tell the story pretty nicely where/why the issue happened - but I've added some ifs